src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -88,26 +106,110 @@ class DocumentStatisticsAnnotationData {
+
+    using FrequencyAndIndexMap              = std::unordered_map<std::string, FrequencyAndIndexStruct>;
     // ----------------------------------------------------------------------
     // |
     // |  Public Data
     // |
     // ----------------------------------------------------------------------
-    FrequencyMap const                      TermFrequency;                  /// Words and the number of documents that it appears in
-    IndexMap const                          TermIndex;
+    FrequencyAndIndexMap const              TermFrequencyAndIndex;
     std::uint32_t const                     TotalNumDocuments;
 
     // ----------------------------------------------------------------------
     // |
     // |  Public Methods
     // |
     // ----------------------------------------------------------------------
-    DocumentStatisticsAnnotationData(FrequencyMap termFreq, IndexMap termIndex, std::uint32_t totalNumDocuments);
+    DocumentStatisticsAnnotationData(FrequencyAndIndexMap termFrequencyAndIndex, std::uint32_t totalNumDocuments);
     ~DocumentStatisticsAnnotationData(void) = default;
 
     FEATURIZER_MOVE_CONSTRUCTOR_ONLY(DocumentStatisticsAnnotationData);
 };
 
+namespace {
+
+inline DocumentStatisticsAnnotationData::FrequencyMap PruneTermFreqMap(DocumentStatisticsAnnotationData::FrequencyMap const & prunedTermFreqByDf, 
+                                                                       std::uint32_t maxFeatures) {
+    std::vector<std::uint32_t> frequencyVector;
+
+    std::transform(prunedTermFreqByDf.begin(), 
+                   prunedTermFreqByDf.end(), 
+                   std::back_inserter(frequencyVector), 
+                   [](auto const & kv) { 
+                       return kv.second; 
+                   });
+
+    //partial sort, ave O(n)
+    std::nth_element(frequencyVector.begin(), 
+                     frequencyVector.begin() + static_cast<int>(maxFeatures) - 1, 
+                     frequencyVector.end(), 
+                     std::greater<std::uint32_t>());
+
+    std::uint32_t targetValue = frequencyVector[maxFeatures - 1];
+    long long remainingTargetValue = std::count(frequencyVector.begin(), 
+                                                frequencyVector.begin() + static_cast<int>(maxFeatures),
+                                                targetValue);
+
+    DocumentStatisticsAnnotationData::FrequencyMap prunedTermFreq;
+    for (auto const & termFrequencyPair : prunedTermFreqByDf) {
+
+        if (termFrequencyPair.second > targetValue)
+            prunedTermFreq.insert(termFrequencyPair);
+
+        if (termFrequencyPair.second == targetValue && remainingTargetValue > 0) {
+            prunedTermFreq.insert(termFrequencyPair);
+            --remainingTargetValue;
+        }
+    }
+
+    return prunedTermFreq;
+}
+
+inline DocumentStatisticsAnnotationData::FrequencyMap PruneTermFreqMap(DocumentStatisticsAnnotationData::FrequencyMap const & termFrequency, 
```

Comment:
I'd suggest that we remove those values that aren't in the desired range so that we aren't building multiple maps.